os/test/src-jvm/ProcessPipelineTests.scala

```diff
@@ -0,0 +1,163 @@
+package test.os
+
+import java.io._
+import java.nio.charset.StandardCharsets
+
+import os._
+import utest._
+import TestUtil.prep
+import scala.util.Try
+
+object ProcessPipelineTests extends TestSuite {
+  val scriptFolder = pwd / "os" / "test" / "resources" / "scripts"
+
+  lazy val scalaHome = sys.env("SCALA_HOME")
+
+  def isWindows = System.getProperty("os.name").toLowerCase().contains("windows")
+
+  def scriptProc(name: String, args: String*): Seq[String] =
+    Seq(
+      "scala-cli" + (if (isWindows) ".bat" else ""),
```

Comment:
Is it possible to use a smaller more ubiquitous process in these test cases here? I don't want to have to install the whole of `scala-cli` just to run unit tests locally, especially when it seems some combination of simple unix tools (`cat`, `sleep`, `bash -c`, etc.) should be able to exercise it

Reply:
lets go with test jars then. Mill makes submodules cheap, we can easily have a few submodules to create jars for testing purposes

Reply:
Could you please provide an example of a decent simple workflow generating jars like this? I am not familiar with Mill, the configuration of the project you provided is quite complex, and my version will probably be a bit overcomplicated, leading to more work in the review-fixes cycle.

Reply:
Something like this should work

```diff
diff --git a/build.sc b/build.sc
index 80437ae..c28ab6b 100644
--- a/build.sc
+++ b/build.sc
@@ -58,6 +58,10 @@ trait MiMaChecks extends Mima {
   )
 }
 
+object testJar1 extends JavaModule
+object testJar2 extends JavaModule
+object testJar3 extends JavaModule
+
 trait OsLibModule
     extends CrossScalaModule
     with PublishModule
@@ -84,7 +88,12 @@ trait OsLibModule
     def ivyDeps = Agg(Deps.utest, Deps.sourcecode)
 
     // we check the textual output of system commands and expect it in english
-    def forkEnv = super.forkEnv() ++ Map("LC_ALL" -> "C")
+    def forkEnv = super.forkEnv() ++ Map(
+      "LC_ALL" -> "C",
+      "TEST_JAR_1_ASSEMBLY" -> testJar1.assembly().path.toString,
+      "TEST_JAR_2_ASSEMBLY" -> testJar2.assembly().path.toString,
+      "TEST_JAR_3_ASSEMBLY" -> testJar3.assembly().path.toString,
+    )
   }
 }
```

You can then put the source code for your test jars in `testJar1/src/TestJar1.java`, and use the assemblies in your test suite via `os.proc("java", "-jar", sys.env("TEST_JAR_1_ASSEMBLY"))).call()`

Reply:
in the tests `java.util.NoSuchElementException: key not found: TEST_JAR_READER_ASEEMBLY` happens for some reason